Index: nmea0183/NMEA0183Msg.cpp
===================================================================
--- nmea0183.orig/NMEA0183Msg.cpp
+++ nmea0183/NMEA0183Msg.cpp
@@ -153,7 +153,7 @@ bool tNMEA0183Msg::Init(const char *_Mes
   Clear();
   size_t nSender=2;
   size_t nMessageCode=0;
-  if ( _Sender==0 && (nSender=strlen(_Sender))>7 ) return false;
+  if ( _Sender==0 || (nSender=strlen(_Sender))>7 ) return false;
   if ( _MessageCode==0 || (nMessageCode=strlen(_MessageCode))>10 ) return false;
 
   Prefix=_Prefix;
Index: nmea0183/NMEA0183.cpp
===================================================================
--- nmea0183.orig/NMEA0183.cpp
+++ nmea0183/NMEA0183.cpp
@@ -185,7 +185,7 @@ bool tNMEA0183::SendBuf(const char *buf)
 
   for (size_t temp = (MsgOutWritePos + 1) % MsgOutBufSize;
        buf[iBuf]!=0 && temp!=MsgOutReadPos;
-       temp = (MsgOutWritePos + 1) % MsgOutBufSize ) {
+       temp = (MsgOutWritePos + 1) % MsgOutBufSize, iBuf++ ) {
     MsgOutBuf[MsgOutWritePos]=buf[iBuf];
     MsgOutWritePos=temp;
   }
Index: nmea0183/NMEA0183Msg.h
===================================================================
--- nmea0183.orig/NMEA0183Msg.h
+++ nmea0183/NMEA0183Msg.h
@@ -95,7 +95,7 @@ class tNMEA0183Msg
     static inline int GetYear(const tmElements_t &TimeElements) { return TimeElements.tm_year+1900; }
     static inline int GetMonth(const tmElements_t &TimeElements) { return TimeElements.tm_mon+1; }
     static inline int GetDay(const tmElements_t &TimeElements) { return TimeElements.tm_mday; }
-    static inline time_t makeTime(tmElements_t &TimeElements) { return mktime(&TimeElements); }
+    static inline time_t makeTime(tmElements_t &TimeElements) { return timegm(&TimeElements); }
     static inline void breakTime(time_t time, tmElements_t &TimeElements) { TimeElements=*localtime(&time); }
     static time_t daysToTime_t(unsigned long val);
     #endif
